### ISSUE TITLE
Use potentially unnecessary use of useEffect in usePrevious

### ISSUE THREAD

**Issue by RickHuizing01** [streamich/react-use#2605]
**Is your feature request related to a problem? Please describe.**

The current implementation of the `usePrevious` hook relies on `useEffect` to store the previous state. While this approach works,  I believe that the use of useEffect might not be necessary. The usage of useEffect could be considered an anti-pattern according to the official React documentation (as described in [You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect)) if I interpret it correctly. I think the `useEffect` in this hook adds unnecessary complexity and potentially increases rendering overhead.

It could be that more hooks such as `usePreviousDistinct` can be refactored without useEffect but I stumbled on this one first so I decided to put my focus this proposal first.

**Describe the solution you'd like**

I propose an alternative implementation of `usePrevious` that eliminates the need for `useEffect` by using only `useRef`. This approach ensures that the hook is purely synchronous and avoids unnecessary side effects. Here's the proposed implementation:

```javascript
import { useRef } from "react";

export const usePrevious = <T>(state: T): T | undefined => {
    const ref = useRef<T>();

    const previousValue = ref.current;
    ref.current = state;

    return previousValue;
};
```
The implementation above updates the `ref` value synchronously with the component's render cycle, without requiring a separate side effect to handle state updates.

I was wondering if the implementation above could provide the same functionality but without using useEffect. It could be that I am missing edge cases but I'm more than happy to hear about them and add changes if possible.

**Describe alternatives you've considered**

None yet but I am open to investigate if necessary.

I would be happy to open a merge request with the proposed changes if this helps.

**Comment by childrentime** [comment]
how about this? https://github.com/alibaba/hooks/blob/master/packages/hooks/src/usePrevious/index.ts

**Comment by FleetAdmiralJakob** [comment]
btw. alibaba has the same discussion 😂: https://github.com/alibaba/hooks/issues/2162

**Comment by FleetAdmiralJakob** [comment]
Yes, totally agree with you @RickHuizing01. For https://github.com/childrentime/reactuse we went with the version without deep equality too: https://github.com/childrentime/reactuse/blob/main/packages/core/src/usePrevious/index.ts

**Comment by qwelol** [comment]
What do you think about this?
https://github.com/streamich/react-use/pull/2617#issuecomment-4864636639
